nav2_behaviors/plugins/assisted_teleop.cpp

```diff
@@ -0,0 +1,186 @@
+    joystick_topic_, rclcpp::SystemDefaultsQoS(),
+    std::bind(
+      &AssistedTeleop::joyCallback,
+      this, std::placeholders::_1));
+}
+
+Status AssistedTeleop::onRun(const std::shared_ptr<const AssistedTeleopAction::Goal> command)
+{
+  command_time_allowance_ = command->time_allowance;
+  end_time_ = steady_clock_.now() + command_time_allowance_;
+  return Status::SUCCEEDED;
+}
+
+void AssistedTeleop::onActionCompletion()
+{
+  RCLCPP_INFO(logger_, "Action completed");
+  input_twist_ = geometry_msgs::msg::Twist();
+  joy_ = sensor_msgs::msg::Joy();
+}
+
+Status AssistedTeleop::onCycleUpdate()
+{
+  rclcpp::Duration time_remaining = end_time_ - steady_clock_.now();
+  if (time_remaining.seconds() < 0.0 && command_time_allowance_.seconds() > 0.0) {
+    stopRobot();
+    RCLCPP_WARN(
+      logger_,
+      "Exceeded time allowance before reaching the BackUp goal - Exiting BackUp");
+    return Status::FAILED;
+  }
+
+  geometry_msgs::msg::PoseStamped current_pose;
+  if (!nav2_util::getCurrentPose(
+      current_pose, *tf_, global_frame_, robot_base_frame_,
+      transform_tolerance_))
+  {
+    RCLCPP_ERROR(logger_, "Current robot pose is not available.");
+    return Status::FAILED;
+  }
+
+  // user states that teleop was successful
+  if (joy_.buttons.size() > 0 && joy_.buttons[0] > 0) {
+    stopRobot();
+    return Status::SUCCEEDED;
+  }
+
+  // user states that teleop failed
+  if (joy_.buttons.size() > 1 && joy_.buttons[1] > 0) {
+    stopRobot();
+    return Status::FAILED;
+  }
+
+  geometry_msgs::msg::Twist cmd_vel;
+  cmd_vel = computeVelocity(input_twist_);
+  vel_pub_->publish(std::move(cmd_vel));
+
+  return Status::RUNNING;
+}
+
+geometry_msgs::msg::Twist AssistedTeleop::computeVelocity(geometry_msgs::msg::Twist & twist)
+{
+  geometry_msgs::msg::PoseStamped current_pose;
+  if (!nav2_util::getCurrentPose(
+      current_pose, *tf_, global_frame_, robot_base_frame_,
+      transform_tolerance_))
+  {
+    RCLCPP_ERROR(logger_, "Current robot pose is not available.");
+    return geometry_msgs::msg::Twist();
+  }
+  geometry_msgs::msg::Pose2D projected_pose;
+  projected_pose.x = current_pose.pose.position.x;
+  projected_pose.y = current_pose.pose.position.y;
+  projected_pose.theta = tf2::getYaw(current_pose.pose.orientation);
+
+  const double dt = 0.01;
+  geometry_msgs::msg::Twist scaled_twist = twist;
+  for (double time = dt; time < projection_time_; time+=dt)
+  {
+    projected_pose = projectPose(projected_pose, twist, dt);
+
+    if (!collision_checker_->isCollisionFree(projected_pose))
+    {
+      RCLCPP_WARN(logger_, "Collision approaching in %.2f seconds", time);
+      double scale_factor = time / projection_time_;
+      scaled_twist.linear.x *= scale_factor;
+      scaled_twist.linear.y *= scale_factor;
+      scaled_twist.angular.z *= scale_factor;
+      break;
+    }
+  }
+  return scaled_twist;
+}
+
+geometry_msgs::msg::Pose2D AssistedTeleop::projectPose(
+  geometry_msgs::msg::Pose2D pose,
```

Comment:
If you made these references `&`, then the signature can be `void` and just update the input of `pose` with the new value implicitly.

Also, `twist` and `projection_time` should be `const`

Reply:
Would this just be to remove the coping of geometry_msgs::msg::Pose2D? Or is there any other gains by passing in the output by reference?

I tend to follow https://www.fluentcpp.com/2016/11/22/make-your-functions-functional/ but really don't mind changing it :)